contracts/AJMarket.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+
+/// @custom:security-contact security@ajcoin.xyz
+contract AJMarket is Ownable {
+    using Counters for Counters.Counter;
+
+    IERC20 ajCoin;
+
+    // Total count of all listings
+    Counters.Counter private _listingCount;
+
+    // Mapping from listing ID to supply
+    mapping(uint256 => uint256) private _listingSupply;
+
+    // Mapping from listing ID to price
+    mapping(uint256 => uint256) private _listingPrice;
+
+    // Total count of all orders
+    Counters.Counter private _orderCount;
+
+    // Mapping from order ID to listing ID
+    mapping(uint256 => uint256) private _orderListing;
+
+    // Mapping from order ID to user address
+    mapping(uint256 => address) private _orderUser;
+
+    // Mapping from order ID to price
+    mapping(uint256 => uint256) private _orderPrice;
+
+    // Mapping user address to order count
+    mapping(address => uint256) private _orderCountForUser;
+
+    // Mapping from user to list of order IDs
+    mapping(address => mapping(uint256 => uint256)) private _orders;
+
+    event Purchase(address indexed user, uint256 indexed listingId, uint256 price);
+
+    constructor(address _ajCoin) {
+        ajCoin = IERC20(_ajCoin);
+    }
+
+    function createListing(uint256 supply, uint256 price) external onlyOwner {
+        uint256 listingId = _listingCount.current();
+        _listingCount.increment();
+    
+        _listingSupply[listingId] = supply;
+        _listingPrice[listingId] = price;
+    }
+
+    function setListingSupply(uint256 listingId, uint256 supply) external onlyOwner {
+        _listingSupply[listingId] = supply;
+    }
+
+    function setListingPrice(uint256 listingId, uint256 price) external onlyOwner {
+        _listingPrice[listingId] = price;
+    }
+
+    function getListing(uint256 listingId) external view returns (uint256, uint256) {
+        return (_listingSupply[listingId], _listingPrice[listingId]);
+    }
+
+    function purchase(uint256 listingId) external {
```

Comment:
so.. im a bit confused on this point.  this is kinda exactly what I was concerned about... but... Did I mitigate enough by having the payment happen first and the call to external contract be a "trusted" contract?

What is the cost of adding the `nonReentrant` modifier?  is it something to worry about as far as gas savings or not really?

Reply:
The ordering of payment happening first and then updating the storage values for listings and indexes by users is valid. Re-entrancy can be an unintended consequence. A simpler way to guard this is to add a global variable to set to "unlocked" by default. When the `purchase` function is invoked, `require(!locked, "Potential reentrancy!");` can guard the lock variable. Within the scope of the function, the locked can be set to true, and then unset before exiting the function.

Reply:
Ahhhh. ok.  read the article again too.  I think I fully get it.  Since this is a trusted contract and ajcoin doesn't call any other contracts on transfer I won't add the guard here for whatever little gas savings it might give me.